rust/headless-client/src/windows.rs

```diff
@@ -152,25 +165,38 @@ fn fallible_windows_service_run(arguments: Vec<OsString>) -> Result<()> {
             // We cancelled because Windows asked us to shut down.
             Ok(())
         }
-        Err(join_error) => {
-            // The IPC task may have panicked
-            Err(join_error.into())
-        }
-        Ok(Err(error)) => Err(error.context("ipc_listen failed")),
+        Err(join_error) => Err(anyhow::Error::from(join_error).context("`ipc_listen` panicked")),
+        Ok(Err(error)) => Err(error.context("`ipc_listen` threw an error")),
         Ok(Ok(impossible)) => match impossible {},
     };
+    if let Err(error) = &result {
+        tracing::error!(?error, "`ipc_listen` failed");
+    }
+
+    // Drop the logging handle so it flushes the logs before we let Windows kill our process.
+    // There is no obvious and elegant way to do this, since the logging and `ServiceState`
+    // changes are interleaved, not nested:
+    // - Start logging
+    // - ServiceState::Running
+    // - Stop logging
+    // - ServiceState::Stopped
+    std::mem::drop(logging_handle);
 
     // Tell Windows that we're stopping
-    status_handle.set_service_status(ServiceStatus {
-        service_type: SERVICE_TYPE,
-        current_state: ServiceState::Stopped,
-        controls_accepted: ServiceControlAccept::empty(),
-        exit_code: ServiceExitCode::Win32(if result.is_ok() { 0 } else { 1 }),
-        checkpoint: 0,
-        wait_hint: Duration::default(),
-        process_id: None,
-    })?;
-    result
+    // Per Windows docs, this will cause Windows to kill our process eventually.
+    status_handle
+        .set_service_status(ServiceStatus {
+            service_type: SERVICE_TYPE,
+            current_state: ServiceState::Stopped,
+            controls_accepted: ServiceControlAccept::empty(),
+            exit_code: ServiceExitCode::Win32(if result.is_ok() { 0 } else { 1 }),
+            checkpoint: 0,
+            wait_hint: Duration::default(),
+            process_id: None,
+        })
+        .expect("Should be able to tell Windows we're stopping");
+    // Generally unreachable
```

Comment:
is there any case when this is reachable? can we add `unreachable!()` and have the return type be `Result<Infallible>`?

Reply:
I don't think we can, we could do something like an infinite sleep or infinite await.
The docs say this https://learn.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-setservicestatus

> Do not attempt to perform any additional work after calling SetServiceStatus with SERVICE_STOPPED, because the service process can be terminated at any time.

I interpret that as, the function will return, and the process might get scheduled again, but then it'll just be killed at some point. So I don't want it to panic if Windows doesn't kill us synchronously. That function seems under-specified

Reply:
gotcha, better safe here and not panic